### ISSUE TITLE
kurento is not accepting ws connections after some hours of work (restar is needed)

### ISSUE THREAD

**Issue by darchuk** [Kurento/bugtracker#259]
**KMS Version**:

Paste output of `kurento-media-server -v` in the machine where KMS is running

**Other libraries versions**:

Paste output of executing `dpkg -l | egrep -i "kurento|gst.*1.5|nice"` in the
machine where KMS is running

**Client libraries**
- Language: <Java, Node.js, JavaScript, homebrewed>
- Version:

**Browsers tested**
  Add OK or FAIL, along with the version, after browsers where you
have tested this issue:
- Chrome:
- Firefox:
- Native:

**System description:**
	Please describe your setup (where is KMS located, where are the
	clients, STUN, TURN...)

**What steps will reproduce the problem?**
1.
2.
3.

**What is the expected result?**

**What happens instead?**

**Does it happen with one of the tutorials?**

**Please provide any additional information below.**

**Comment by xenyou** [comment]
I have the same problem. very critical!!

**Comment by xenyou** [comment]
Although I don't know if it's related to this, garbage collector often run just before stalling KMS.

2018-07-12 15:00:06,045955 24956 [0x00007fed74fd8700]   debug KurentoMediaSet           MediaSet.cpp:122 doGarbageCollection()  Running garbage collector

Does gc something evil?

**Comment by zhangalex** [comment]
I think you'd better pay for commercial support if it is emergency.

**Comment by PonyHugger** [comment]
Same problem over here!
We need to get this fixed!

**Comment by Jaehdawg** [comment]
We see this issue as well.

**Comment by buzeandrei** [comment]
We are having the same issue with a client in production. After a couple of hours of streaming Kurento doesn't accept connections anymore and we need to restart the app. We think it might be because of insufficient CPU/RAM as after it crashes almost 100% of the memory is used.

**Comment by micaelgallego** [comment]
What kind of connections KMS is not attending? WebSocket connections using KurentoClient?

**Comment by darchuk** [comment]
> What kind of connections KMS is not attending? WebSocket connections using KurentoClient?
Correct.
btw, I can't even open WebSocket connection with kurento directly, bypassing kurentoClient

**Comment by buzeandrei** [comment]
In my case, Kurento seemed not to listen on the default WebSocket port at all.

**Comment by micaelgallego** [comment]
Usually Application Server uses KurentoClient to connect to KMS and maintain WebSocket connection opened. Is this your case? Requests give you timeouts? Or you connect and disconnect KurentoClient from time to time?

We are preparing KMS load testing to try to solve these issues.

**Comment by darchuk** [comment]
Yes, we use KurentoClient to maintain WebSocket connections, and keep KurentoClient connected as long as application server is running.
At some time kurento is starting to send timeout exceptions (we use default 30s timeout)

**Comment by xenyou** [comment]
There are two cases in our production.

### case1
1. boot kurento-media-server
2. connect to the server using WebSocket and the connection is established.
3. Do some conversation with KMS via kurento-client (node module)
4. at some time, kurento-client got timeout for some request.

### case2
1. boot kurento-media-server
2. try to connect to the WebSocket, but can not connect.
kurento-client module says this error message repeatedly.

```
reconnect to server 1549 10000 dd6db11f-ab91-4fd5-b0d2-3bb80504905f
reconnect to server 1549 10000 a9a72208-52df-4e32-ac1f-8b96fac683e8
reconnect to server 1549 10000 b1ef16c9-72f3-4db5-b6f8-9ba5d3482cdf
reconnect to server 1549 10000 75c19c81-bc60-457d-99e2-5c122a349986
```

**Comment by xenyou** [comment]
Because of the case2, I don't think that load is so important.

**Comment by alexei-talankov-oxagile** [comment]
Any updates on this?

**Comment by j1elo** [comment]
Please change [this line](https://github.com/Kurento/kms-core/blob/eb4bd3a3d0e0919fc33ecbfcf3c97e56d43ff4ec/src/server/implementation/EventHandler.cpp#L27) to increment the number of threads from `1` to `10`:

```c
static WorkerPool workers (10);
```

then build and test if this improves the issue (instructions here to [build from sources](https://doc-kurento.readthedocs.io/en/latest/dev/dev_guide.html#building-from-sources))

**Comment by j1elo** [comment]
You can now test this with an experimental installation of Kurento, using this apt-get repository in your `/etc/apt/sources.list.d/kurento.list`:

```
deb http://ubuntu.openvidu.io/eventhandler-10-threads xenial kms6
```

Or with this Docker image https://hub.docker.com/r/kurento/kurento-media-server-exp
```
kurento/kurento-media-server-exp:eventhandler-10-threads
```

**Comment by darchuk** [comment]
> You can now test this with an experimental installation of Kurento, using this apt-get repository in your `/etc/apt/sources.list.d/kurento.list`:
> 
> ```
> deb http://ubuntu.openvidu.io/eventhandler-10-threads xenial kms6
> ```
> 
> Or with this Docker image https://hub.docker.com/r/kurento/kurento-media-server-exp
> 
> ```
> kurento/kurento-media-server-exp:eventhandler-10-threads
> ```

Hi j1elo, thanks for answer.
Why should we prefer experimental kurento over usual one, what if we just fork kurento module and change it by ourselves?

**Comment by j1elo** [comment]
It's just a more convenient method for some users that don't want or don't know how to build from sources; I had to do the images for some internal tests anyway, so I just told you about them here.

Of course, if you can just do as I said in my previous comment https://github.com/Kurento/bugtracker/issues/259#issuecomment-523077731 then just do it and let me know if the change improves anything

**Comment by xenyou** [comment]
Using KMS 6.10 with Ubuntu 18.04 might solve this issue.

We recently changed the versions (KMS 6.10 from 6.6, Ubuntu 18.04 from 16.04).
And the issue has not been happened yet.

But I'm not sure. I'll watch the situation for a while.

**Comment by xenyou** [comment]
It's not happened yet after upgrading Ubuntu and KMS.

**Comment by hdev72** [comment]
I have this problem too, any body resolved this?
